src/wallet/wallet.cpp

```diff
@@ -7165,13 +7212,16 @@ std::optional<libzcash::UnifiedAddress> UnifiedAddressForReceiver::operator()(co
         const auto& metadata = wallet.mapUfvkAddressMetadata.at(ufvkid);
         auto saplingKey = ufvk.value().GetSaplingKey();
         if (saplingKey.has_value()) {
-            diversifier_index_t j = saplingKey.value().DecryptDiversifier(saplingAddr.d);
-            auto receivers = metadata.GetReceivers(j);
-            if (receivers.has_value()) {
-                auto addr = ufvk.value().Address(j, receivers.value());
-                auto addrPtr = std::get_if<std::pair<UnifiedAddress, diversifier_index_t>>(&addr);
-                if (addrPtr != nullptr) {
-                    return addrPtr->first;
+            auto j = saplingKey.value().DecryptDiversifier(saplingAddr);
+            if (j.has_value()) {
+                auto receivers = metadata.GetReceivers(j.value().first);
+                // Only return a unified address for external addresses
```

Comment:
What does the caller of `UnifiedAddressForReceiver` expect `std::nullopt` to mean? That we could not find a UA corresponding to this receiver? Or that we could, and it is a change address that should not be displayed? I suspect the former is what is currently assumed, which would cause that code to assume it is a legacy Sapling address to be shown directly, instead of a known UA change address that should not.